runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/ResponseLengthValidationInterceptor.kt

```diff
@@ -0,0 +1,84 @@
+    override suspend fun modifyBeforeDeserialization(context: ProtocolResponseInterceptorContext<Any, HttpRequest, HttpResponse>): HttpResponse {
+        val response = context.protocolResponse
+
+        return response.headers["Content-Length"]?.let {
+            response.copy(body = response.body.toLengthValidatingBody(it.toLong()))
+        } ?: response
+    }
+}
+
+@InternalApi
+private fun HttpBody.toLengthValidatingBody(expectedContentLength: Long): HttpBody = when (this) {
+    is HttpBody.SourceContent -> LengthValidatingSource(readFrom(), expectedContentLength).toHttpBody(contentLength)
+    is HttpBody.ChannelContent -> LengthValidatingByteReadChannel(readFrom(), expectedContentLength).toHttpBody(contentLength)
+    else -> throw ClientException("HttpBody type is not supported")
+}
+
+/**
+ * An [SdkSource] which keeps track of how many bytes were consumed. After the underlying source is exhausted or too many
+ * bytes are read, it compares the number of bytes received against the expected length.
+ * @param source The underlying [SdkSource] to read from
+ * @param expectedContentLength The expected content length as described by the `Content-Length` header
+ * @throws [IllegalStateException] if there is a mismatch.
+ */
+private class LengthValidatingSource(
+    private val source: SdkSource,
+    private val expectedContentLength: Long,
+) : SdkSource by source {
+    var bytesReceived = 0L
+    override fun read(sink: SdkBuffer, limit: Long): Long = source.read(sink, limit).also {
+        if (it == -1L || bytesReceived > expectedContentLength) {
+            validateContentLength(expectedContentLength, bytesReceived)
+        } else {
+            bytesReceived += it
+        }
+    }
+}
+
+/**
+ * An [SdkByteReadChannel] which keeps track of how many bytes were consumed. After the underlying channel is exhausted
+ * or too many bytes are read, it compares the number of bytes received against the expected length.
+ * @param chan The underlying [SdkByteReadChannel] to read from
+ * @param expectedContentLength The expected content length as described by the `Content-Length` header
+ * @throws [IllegalStateException] if there is a mismatch.
+ */
+private class LengthValidatingByteReadChannel(
+    private val chan: SdkByteReadChannel,
+    private val expectedContentLength: Long,
+) : SdkByteReadChannel by chan {
+    var bytesReceived = 0L
+
+    override suspend fun read(sink: SdkBuffer, limit: Long): Long = chan.read(sink, limit).also {
+        if (chan.isClosedForRead || bytesReceived > expectedContentLength) {
+            if (it != -1L) { bytesReceived += it }
+            validateContentLength(expectedContentLength, bytesReceived)
+        } else {
+            bytesReceived += it
+        }
+    }
+}
+
+private fun validateContentLength(expected: Long, actual: Long) {
+    check(expected == actual) { "Total bytes consumed ($actual) does not match expected ($expected)." }
+}
```

Comment:
**Correctness**: This error seems a little opaque for users. While I think the actual/expected byte counts are interesting, we should probably mention that this is likely because a connection closed prematurely or something like that. We should probably also throw an `IOException` (or something similar) instead of `IllegalStateException`.

Reply:
We have an [EOFException](https://github.com/awslabs/smithy-kotlin/blob/71db9ef7a8cfd28d79ee2d542d62827d1d3e6a17/runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/io/Exceptions.kt#L12) defined in the runtime that inherits from IOException